### PR TITLE
Append a new runner name for the custom object container

### PR DESCRIPTION
**Pull request by JudeNiroshan**
- Define new optional alphanumeric (up to 20 chars - the limitation of API) CLI parameter called "runnerName" which should be unique for combination of CLI runner of same resource and target CTP project
- In case "runnerName" **is not set** the application should create custom objects with following containers:
" commercetools-project-sync.runnerName. {resource} .timestampGenerator "
" commercetools-project-sync.runnerName.typeSync "
- In case "runnerName" is set:
" commercetools-project-sync. {runnerName}. {resource} .timestampGenerator "
" commercetools-project-sync. {runnerName} .typeSync "

### CODE REVIEW

File: src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java [hunk 1]
```diff
@@ -90,11 +99,28 @@ public CustomObjectServiceImpl(@Nonnull final SphereClient sphereClient) {
     return createCustomObject(lastSyncCustomObjectDraft);
   }
 
+  @Override
+  public CustomObjectService attachRunnerName(String runnerName) {
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java]: Please use `final` for the parameter. Also please use `@Nonnull` or `@Nullable` as it makes it easier to know if it could be passed as `null` or not.

File: src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java [hunk 2]
```diff
@@ -90,11 +99,28 @@ public CustomObjectServiceImpl(@Nonnull final SphereClient sphereClient) {
...
+    if (runnerName != null && !runnerName.isEmpty()) {
+      this.runnerName = runnerName;
+    }
+    return this;
+  }
+
+  @Override
+  public CustomObjectService attachMethodName(String methodName) {
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java]: Please use `final` for the parameter. Also please use `@Nonnull` or `@Nullable` as it makes it easier to know if it could be passed as `null` or not.

File: src/main/java/com/commercetools/project/sync/Syncer.java [hunk 3]
```diff
@@ -114,6 +114,8 @@ public Syncer(
     }
 
     return customObjectService
+        .attachRunnerName(runnerName)
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/project/sync/Syncer.java]: There is a problem with this design -> Now the service is stateful rather stateless. Now each service is bound to one `runnerName` and one `syncModuleName`. But what if we want to get the `getCurrentCtpTimestamp()` of  another runnerName of another sync module? then we always have to attach first?

this is as opposed to `getCurrentCtpTimestamp(runnerName, syncModuleName)` where the service is stateless and only depends on the input given not some predefined value. Do you know what I mean?

File: src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java [hunk 4]
```diff
@@ -28,6 +30,8 @@
   private static final long MINUTES_BEFORE_CURRENT_TIMESTAMP = 2;
 
   private SphereClient sphereClient;
+  private String runnerName = DEFAULT_RUNNER_NAME;
+  private String methodName = StringUtils.EMPTY;
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java]: I don't mind personally either `""` or  `StringUtils.EMPTY`. But what I think we should do is be consistent. So use only of them so it's not confusing why sometimes we use this and why the other. 

https://github.com/commercetools/commercetools-project-sync/pull/30/files#diff-90d173bf2f68b44092d0e0fa23e07f36R29 we use ""

File: src/test/java/com/commercetools/project/sync/SyncerFactoryTest.java [hunk 5]
```diff
@@ -52,7 +54,7 @@
   private static final TestLogger cliRunnerTestLogger =
       TestLoggerFactory.getTestLogger(CliRunner.class);
 
-  @AfterEach
+  @BeforeEach
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/project/sync/SyncerFactoryTest.java]: why is this changed?

File: src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java [hunk 6]
```diff
@@ -67,12 +67,13 @@
 import uk.org.lidalia.slf4jtest.TestLogger;
 import uk.org.lidalia.slf4jtest.TestLoggerFactory;
 
-class CliRunnerIT {
+class MOCK_RUNNER_NAME {
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java]: yes, It's a typo. I have reverted it back

File: README.md [hunk 7]
```diff
@@ -21,6 +21,7 @@ Dockerized CLI application which allows to automatically sync different resource
 ### Prerequisites
  
  - Make sure you have `JDK 8` installed.
+ - Intellij IDEA version 2018.2 or higher version
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply README.md]: 5b96e1c

File: src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java]: 58ef1fa4239d90216da9510f3d0f242b78b90542

[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java]: 58ef1fa4239d90216da9510f3d0f242b78b90542

File: README.md [hunk 8]
```diff
@@ -47,9 +48,10 @@ Dockerized CLI application which allows to automatically sync different resource
 
    ```bash
    usage: commercetools-project-sync
-    -h,--help         Print help information.
-    -s,--sync <arg>   Choose one of the following modules to run: "types", "productTypes", "categories", "products", "inventoryEntries" or "all" (will run all the modules).
-    -v,--version      Print the version of the application.
+    -h,--help               Print help information.
+    -s,--sync <arg>         Choose one of the following modules to run: "types", "productTypes", "categories", "products", "inventoryEntries" or "all" (will run all the modules).
+    -r,--runnerName <arg>   Unique identifer name to run parallel sync in same project. (optional parameter)
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply README.md]: 4609942b315cd888c049dc6eed6ebeaa4a26a4fd

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply README.md]: 5b96e1caa76cbda67ba98ab62739c08288d75bbf

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply README.md]: 821ee40a077525f25041a047affbf44a8af25f7b
4609942b315cd888c049dc6eed6ebeaa4a26a4fd

File: README.md [hunk 9]
```diff
@@ -107,6 +109,11 @@ commercetools/commercetools-project-sync:1.1.0 -s all
 - To run the inventoryEntry sync
    ```bash
    docker run commercetools/commercetools-project-sync:1.1.0 -s inventoryEntries
+   ```   
+       
+- To run the all sync with unique runner name
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply README.md]: 6f0173b67790bb9c54bff9366a36465ddb138813

File: src/main/java/com/commercetools/project/sync/SyncerFactory.java [hunk 10]
```diff
@@ -55,20 +55,27 @@ public static SyncerFactory of(
   }
 
   @Nonnull
-  CompletableFuture<Void> syncAll() {
+  CompletableFuture<Void> syncAll(String runnerName) {
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/project/sync/SyncerFactory.java]: d885f608dd0665dfba5373eff4584aa61a3765bd

File: src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java (hunk 1 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java]: a5b7531c545f6f905e4e58d8bab4838fab0c8eea

File: src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java (hunk 2 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java]: a5b7531c545f6f905e4e58d8bab4838fab0c8eea

File: src/main/java/com/commercetools/project/sync/Syncer.java (hunk 3 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/project/sync/Syncer.java]: f71d1a52f166cb28c80d977bc03438630c538a9c

File: src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java (hunk 4 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/project/sync/service/impl/CustomObjectServiceImpl.java]: 6061a6b87f32a769d39e69230a7f08217afd373b

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply README.md]: but I still don't get why is this a note even?

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply README.md]: 16f57ee

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply README.md]: 6e63a46

**Comment by heshamMassoud** [comment]
Looks good. Thanks @JudeNiroshan 🚀

**Comment by ahmetoz** [comment]
good job 👍
